### PR TITLE
Move log sanitization to implementable

### PR DESCRIPTION
**Pull request by serathius**
/cc @44past4 @dashpole @logicalhan @wojtek-t @ehashman @brancz

### CODE REVIEW

**Comment by dashpole** [comment]
/approve
/hold
After @wojtek-t finishes the PRR, feel free to remove the hold.

Did you get any input from sig-testing on the test plans?  If not, it would be good to have someone ack them.

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 1]
```diff
@@ -198,6 +204,16 @@ To allow configuring if logs should be sanitized we will introduce a new logging
 
 ### Test Plan
 
+Tests should cover two things:
+* Test feature log sanitization works e2e
+* Test Kubernetes components don't leak sensitive data
```
Comment (ehashman) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:209]: What about any potential performance degradations?

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 2]
```diff
@@ -246,3 +265,226 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+
+### Rollout, Upgrade and Rollback Planning
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can a rollout fail? Can it impact already running workloads?**
+  Try to be as paranoid as possible - e.g., what if some components will restart
+   mid-rollout?
+
+   N/A
```
Comment (ehashman) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:335]: I disagree that this is not applicable.

One pitfall I could anticipate is if there is a bug in log sanitization that causes log entries to be mangled and hence compromises the integrity of log data. I'd say it's a low risk that's addressed by having the command line flag to disable it, but I think the risk is there.

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 3]
```diff
@@ -246,3 +265,226 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+
+* **What specific metrics should inform a rollback?**
```
Comment (ehashman) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:337]: Loss of log integrity.

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 4]
```diff
@@ -246,3 +265,226 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+
+* **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
+  Describe manual testing that was done and the outcomes.
+  Longer term, we may want to require automated upgrade/rollback tests, but we
+  are missing a bunch of machinery and tooling and can't do that now.
```
Comment (ehashman) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:342]: Assuming e2e tests pass with both the CLI flag enabled/disabled, I think this is covered.

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 5]
```diff
@@ -246,3 +265,226 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+
+* **Is the rollout accompanied by any deprecations and/or removals of features, APIs,
+fields of API types, flags, etc.?**
+  Even if applying deprecation policies, they may still surprise some users.
```
Comment (ehashman) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:346]: Yes, please mention the CLI flags here.

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 6]
```diff
@@ -246,3 +265,226 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+
+### Monitoring Requirements
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can an operator determine if the feature is in use by workloads?**
+  Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+  checking if there are objects with field X set) may be a last resort. Avoid
+  logs or events for this purpose.
```
Comment (ehashman) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:355]: Command line arguments of running control plane processes.

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 7]
```diff
@@ -246,3 +265,226 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+
+* **What are the SLIs (Service Level Indicators) an operator can use to determine
+the health of the service?**
+  - [ ] Metrics
+    - Metric name:
+    - [Optional] Aggregation method:
+    - Components exposing the metric:
+  - [ ] Other (treat as last resort)
+    - Details:
```
Comment (ehashman) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:364]: N/A

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 8]
```diff
@@ -246,3 +265,226 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+
+* **What are the reasonable SLOs (Service Level Objectives) for the above SLIs?**
+  At a high level, this usually will be in the form of "high percentile of SLI
+  per day <= X". It's impossible to provide comprehensive guidance, but at the very
+  high level (needs more precise definitions) those may be things like:
+  - per-day percentage of API calls finishing with 5XX errors <= 1%
+  - 99% percentile over day of absolute value from (job creation time minus expected
+    job creation time) for cron job <= 10%
+  - 99,9% of /health requests per day finish with 200 code
```
Comment (ehashman) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:373]: N/A

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 9]
```diff
@@ -246,3 +265,226 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+
+* **Are there any missing metrics that would be useful to have to improve observability
+of this feature?**
+  Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
+  implementation difficulties, etc.).
```
Comment (ehashman) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:378]: N/A, this is a logging change.

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 10]
```diff
@@ -246,3 +265,226 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+
+### Dependencies
+
+_This section must be completed when targeting beta graduation to a release._
```
Comment (ehashman) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:382]: Doesn't need to be completed yet as this is in alpha.

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 11]
```diff
@@ -246,3 +265,226 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
+
+  No
+
+* **Will enabling / using this feature result in increasing time taken by any
+operations covered by [existing SLIs/SLOs]?**
+  Think about adding additional work or introducing new steps in between
+  (e.g. need to do X to start a container), etc. Please describe the details.
+
+  No
```
Comment (ehashman) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:450]: If log write time is increased then I'd say this does potentially affect existing SLIs/SLOs, but the likelihood of significant performance degradation is low.

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 12]
```diff
@@ -246,3 +265,226 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+* **Will enabling / using this feature result in non-negligible increase of
+resource usage (CPU, RAM, disk, IO, ...) in any components?**
+  Things to keep in mind include: additional in-memory state, additional
+  non-trivial computations, excessive access to disks (including increased log
+  volume), significant amount of data sent and/or received over network, etc.
+  This through this both in small and large cases, again with respect to the
+  [supported limits].
+
+  Sanitizing logs will require additional computation for performance sensitive logging calls. Feature will be disabled
+  by default until we are able to address them in Beta. More detail in [#Performance overhead] section of KEP.
```
Comment (ehashman) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:461]: :+1:

File: keps/sig-instrumentation/1753-logs-sanitization/kep.yaml [hunk 13]
```diff
@@ -5,9 +5,11 @@ authors:
...
   - "@PurelyApplied"
 owning-sig: sig-instrumentation
-status: provisional
+status: implementable
 creation-date: 2020-05-07
 reviewers:
-  - TBD
 approvers:
-  - TBD
+  - "@dashpole"
```
Comment (ehashman) [inline keps/sig-instrumentation/1753-logs-sanitization/kep.yaml:12]: You can add me as well if you like as either a reviewer or approver.

**Review by ehashman — CHANGES_REQUESTED** [review]
Some minor comments, caught some small bits that are incomplete. Will review again on Monday.

File: keps/sig-instrumentation/1753-logs-sanitization/README.md (hunk 1 above)
Comment (serathius) [reply keps/sig-instrumentation/1753-logs-sanitization/README.md:209]: I don't think we should add specific mechanism to monitor performance degradation and can rely on existing qualification done by scalability team. 

Before enabling this feature by default we will work with SIG Scalability on qualifying (e.g. before major changes for Structured Logging we requested to run scalability tests on PR before merging)

Before feature is enabled by default, performance should not be big concern we should just ensure that e2e periodic test doesn't fail due to overhead

File: keps/sig-instrumentation/1753-logs-sanitization/README.md (hunk 2 above)
Comment (serathius) [reply keps/sig-instrumentation/1753-logs-sanitization/README.md:335]: Ok, I see your point. Maybe let's revisit this during graduation to beta?

File: keps/sig-instrumentation/1753-logs-sanitization/README.md (hunk 11 above)
Comment (serathius) [reply keps/sig-instrumentation/1753-logs-sanitization/README.md:450]: As section below focuses on increased resources there.

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 14]
```diff
@@ -246,3 +265,145 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+
+      Enabling log sanitization will require restarting control plane.
+
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+      Switching between log formats will require restarting all node components
```
Comment (wojtek-t) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:295]: Enabling log sanitization (not swtiching between log formats) :)

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 15]
```diff
@@ -246,3 +265,145 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+  the enablement)?**
+  Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+  Describe the consequences on existing workloads (e.g., if this is a runtime
+  feature, can it break the existing applications?).
+
+  Yes.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+
+  N/A
```
Comment (wojtek-t) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:317]: This one is always applicable :)

The feature will simply start working again :)

File: keps/sig-instrumentation/1753-logs-sanitization/README.md [hunk 16]
```diff
@@ -246,3 +265,145 @@ Static analysis, with its whitebox visibility, is certainly the more thorough ap
...
+
+* **Are there any tests for feature enablement/disablement?**
+  The e2e framework does not currently support enabling or disabling feature
+  gates. However, unit tests in each component dealing with managing data, created
+  with and without the feature, are necessary. At the very least, think about
+  conversion tests if API types are being modified.
+
+  N/A
```
Comment (wojtek-t) [inline keps/sig-instrumentation/1753-logs-sanitization/README.md:325]: Can you add: "We will run manual tests before enabling the feature by default".

File: keps/sig-instrumentation/1753-logs-sanitization/README.md (hunk 11 above)
Comment (wojtek-t) [reply keps/sig-instrumentation/1753-logs-sanitization/README.md:390]: It potentially can for the logs that are synchronous (due to what Elana mentioned) [so error ones, right?]

I agree it's low risk, but I think it's worth mentioning.

**Review by wojtek-t — COMMENTED** [review]
Couple smaller comments.

File: keps/sig-instrumentation/1753-logs-sanitization/README.md (hunk 1 above)
Comment (wojtek-t) [reply keps/sig-instrumentation/1753-logs-sanitization/README.md:209]: It's not a matter of just "not failing" - we shouldn't eat all the slack we have with that. It's a matter of overhead.

But I agree that I wouldn't add a dedicated stuff for it per-se.

**Comment by serathius** [comment]
@dashpole, I reached out to sig-testing on #slack on Friday, but didn't get any response. Will try again

**Comment by serathius** [comment]
/cc @BenTheElder @spiffxp @stevekuznetsov 
PTAL

**Comment by serathius** [comment]
ping @ehashman

**Review by ehashman — APPROVED** [review]
/lgtm

**Comment by ehashman** [comment]
Waiting on SIG Testing to remove hold, this is GTG from the Instrumentation side.

**Review by spiffxp — APPROVED** [review]
/lgtm
as SIG Testing chair

I'm going to assume extensive unit and integration testing comes first, so we're not just relying on e2e tests. I otherwise agree with what I see here as far as test plan and evaluating potential performance impact

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/enhancements/pull/2052#pullrequestreview-502399740 +1

**Comment by lavalamp** [comment]
Did anyone from API Machinery review this? I really think we should make sure this mechanism doesn't collide with things we want to do in the future, and it looks like it might. @deads2k @liggitt

**Comment by lavalamp** [comment]
Also SIG Auth, and someone from the API review/approval process from SIG Arch?

**Comment by lavalamp** [comment]
@apelisse / @sttts can you review this?

**Pull request merged**